File: webui/src/main/scala/com/taskadapter/reporting/ExportResultsFormatter.scala
package com.taskadapter.reporting

import com.taskadapter.web.uiapi.DecodedTaskError
import com.taskadapter.webui.results.ExportResultFormat

import scala.collection.mutable.ListBuffer

object ExportResultsFormatter {
  def toNiceString(result: ExportResultFormat): String = {
    System.lineSeparator() + "Tasks created: " + result.createdTasksNumber + System.lineSeparator() +
      "Tasks updated: " + result.updatedTasksNumber + System.lineSeparator() +
      "General errors: " + result.generalErrors + System.lineSeparator() +
      "Task-specific errors: " + System.lineSeparator() + formatTaskErrors(result.taskErrors)
  }

  def formatTaskErrors(errors: Seq[DecodedTaskError]): String = {
    if (errors.isEmpty) {
      return ""
    }

    var results = new ListBuffer[DecodedTaskError]()
    results += errors.head

    for (i <- 1 until errors.size) {
      val prev = errors(i - 1)
      val current = errors(i)

      val newItem = if (current.exceptionStackTrace == prev.exceptionStackTrace) {
        new DecodedTaskError(current.sourceSystemTaskId, current.connector2ErrorText, "same as previous")
      } else {
        new DecodedTaskError(current.sourceSystemTaskId, current.connector2ErrorText, StacktraceCleaner.stripInternalStacktraceItems(current.exceptionStackTrace))
      }
      results += newItem
    }
    results.map(e => e.sourceSystemTaskId + " - " + e.connector2ErrorText + " - " + e.exceptionStackTrace)
      .mkString(System.lineSeparator())
  }
}

File: webui/src/main/java/com/taskadapter/reporting/ExportResultsFormatter.java
package com.taskadapter.reporting;

import com.taskadapter.web.uiapi.DecodedTaskError;
import com.taskadapter.webui.results.ExportResultFormat;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

public class ExportResultsFormatter {
    public static String toNiceString(ExportResultFormat result) {
        return System.lineSeparator()
                + "Tasks created: " + result.getCreatedTasksNumber() + System.lineSeparator() +
                "Tasks updated: " + result.getUpdatedTasksNumber() + System.lineSeparator() +
                "General errors: " + result.getGeneralErrors() + System.lineSeparator() +
                "Task-specific errors: " + System.lineSeparator() + formatTaskErrors(result.getTaskErrors());
    }

    static String formatTaskErrors(List<DecodedTaskError> errors) {
        if (errors.isEmpty()) {
            return "";
        }

        var results = new ArrayList<DecodedTaskError>();
        results.add(errors.get(0));

        for (int i = 1; i < errors.size(); i++) {
            var prev = errors.get(i - 1);
            var current = errors.get(i);

            var newItem = current.exceptionStackTrace.equals(prev.exceptionStackTrace) ?
                    new DecodedTaskError(current.sourceSystemTaskId, current.connector2ErrorText, "same as previous")
                    :
                    new DecodedTaskError(current.sourceSystemTaskId, current.connector2ErrorText,
                            StacktraceCleaner.stripInternalStacktraceItems(current.exceptionStackTrace));

            results.add(newItem);
        }
        final List<String> list = results.stream().map(e -> e.sourceSystemTaskId + " - " + e.connector2ErrorText + " - " + e.exceptionStackTrace).collect(Collectors.toList());
        return String.join(System.lineSeparator(), list);
    }
}
